### PR TITLE
Jwt

### PR DESCRIPTION
**Pull request by artemgael96**

### CODE REVIEW

File: InternshipBE/ExaLearn.Dal/Database/DbInitializer.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                await SeedRoles(roleManager);
+
+            if (!userManager.Users.Any())
+                await SeedUsers(userManager);
+        }
+
+        public static async Task SeedUsers(UserManager<ApplicationUser> userManager)
+        {
+            var adminEmail = "adminexa@mailnesia.com";
+            await userManager.CreateAsync(new ApplicationUser { UserName = adminEmail, Email = adminEmail, FirstName = "admin", LastName = "Exa" }, "_Test1234");
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Database/DbInitializer.cs]: make normal FirstName LastName

File: InternshipBE/ExaLearn.Dal/Database/ExaLearnDbContext.cs [hunk 2]
```diff
@@ -1,15 +1,19 @@
...
 
-namespace ExaLearn.Dal.Database
-{
-    public class ExaLearnDbContext : DbContext
-    {
-        public ExaLearnDbContext(DbContextOptions<ExaLearnDbContext> options) : base(options)
-        {
-
-        }
+namespace ExaLearn.Dal.Database {
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Database/ExaLearnDbContext.cs]: styling

File: InternshipBE/ExaLearn.Dal/Database/ExaLearnDbContext.cs [hunk 3]
```diff
@@ -1,15 +1,19 @@
...
+    public class ExaLearnDbContext : IdentityDbContext<ApplicationUser, IdentityRole<int>,int> {
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Database/ExaLearnDbContext.cs]: styling

File: InternshipBE/ExaLearn.Dal/Entities/ApplicationUser.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+﻿using Microsoft.AspNetCore.Identity;
+using System.Collections.Generic;
+
+namespace ExaLearn.Dal.Entities {
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Entities/ApplicationUser.cs]: styling

File: InternshipBE/ExaLearn.Dal/Entities/ApplicationUser.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+    public class ApplicationUser : IdentityUser<int> {
+        public string FirstName { get; set; }
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Entities/ApplicationUser.cs:8]: spaces

File: InternshipBE/ExaLearn.Dal/Entities/LoginModel.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace ExaLearn.Dal.Entities {
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Entities/LoginModel.cs]: styling

File: InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+using Microsoft.Extensions.Configuration;
+using Microsoft.IdentityModel.Tokens;
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ExaLearn.WebApi.Controllers {
```
Comment (romanstat) [inline InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs]: styling

File: InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    [Route("api/[controller]")]
+    [ApiController]
+    public class AuthenticateController : ControllerBase {
```
Comment (romanstat) [inline InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs]: space

File: InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        private readonly UserManager<ApplicationUser> userManager;
+        private readonly IConfiguration _configuration;
+
+        public AuthenticateController(UserManager<ApplicationUser> userManager, IConfiguration configuration) {
+            this.userManager = userManager;
+            _configuration = configuration;
+        }
+
+        [HttpPost]
+        [Route("login")]
```
Comment (vlad12345123) [inline InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs]: remove it

File: InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        public async Task<IActionResult> Login([FromBody] LoginModel model) {
+            var user = await userManager.FindByNameAsync(model.Username);
+            if (user != null && await userManager.CheckPasswordAsync(user, model.Password)) {
```
Comment (romanstat) [inline InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs]: space

File: InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+using Microsoft.IdentityModel.Tokens;
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace ExaLearn.WebApi.Controllers {
+    [Route("api/[controller]")]
```
Comment (vlad12345123) [inline InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs]: api/authenticate

File: InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+            var user = await userManager.FindByNameAsync(model.Username);
+            if (user != null && await userManager.CheckPasswordAsync(user, model.Password)) {
+                var userRoles = await userManager.GetRolesAsync(user);
+
+                var authClaims = new List<Claim> {
+                    new Claim(ClaimTypes.Name, user.UserName),
+                    new Claim(JwtRegisteredClaimNames.Jti, Guid.NewGuid().ToString()),
+                };
+
+                foreach (var userRole in userRoles) {
```
Comment (romanstat) [inline InternshipBE/ExaLearn.WebApi/Controllers/AuthenticateController.cs]: each user have 1 role, remove foreach

File: InternshipBE/ExaLearn.WebApi/Startup.cs [hunk 13]
```diff
@@ -27,13 +32,60 @@ public Startup(IConfiguration configuration)
...
+                    In = ParameterLocation.Header,
+                    Description = "Please insert JWT with Bearer into field",
+                    Name = "Authorization",
+                    Type = SecuritySchemeType.ApiKey,
+                    BearerFormat = "JWT",
+                    Scheme = "Bearer"
+                });
+                c.AddSecurityRequirement(new OpenApiSecurityRequirement {
+                {
+                    new OpenApiSecurityScheme
```
Comment (romanstat) [inline InternshipBE/ExaLearn.WebApi/Startup.cs:52]: styling

File: InternshipBE/ExaLearn.WebApi/Startup.cs [hunk 14]
```diff
@@ -27,13 +32,60 @@ public Startup(IConfiguration configuration)
...
+            services.AddIdentity<ApplicationUser, IdentityRole<int>>()
+                    .AddEntityFrameworkStores<ExaLearnDbContext>()
+                    .AddDefaultTokenProviders();
+            services.AddAuthentication(options =>
+            {
+                options.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
+                options.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
+                options.DefaultScheme = JwtBearerDefaults.AuthenticationScheme;
+            })
+
```
Comment (romanstat) [inline InternshipBE/ExaLearn.WebApi/Startup.cs]: spaces

File: InternshipBE/ExaLearn.Dal/Database/ExaLearnDbContext.cs [hunk 15]
```diff
@@ -1,15 +1,21 @@
...
     {
-        public ExaLearnDbContext(DbContextOptions<ExaLearnDbContext> options) : base(options)
+    public class ExaLearnDbContext : IdentityDbContext<ApplicationUser, IdentityRole<int>,int> 
         {
-
-        }
+        public ExaLearnDbContext(DbContextOptions<ExaLearnDbContext> options) : base(options) { }
 
         public DbSet<FileEntry> FileEntry { get; set; }
+        protected override void OnModelCreating(ModelBuilder builder)
```
Comment (romanstat) [inline InternshipBE/ExaLearn.Dal/Database/ExaLearnDbContext.cs:16]: Space

File: InternshipBE/ExaLearn.WebApi/Controllers/DatabaseController.cs [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+using ExaLearn.Dal.Entities;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Mvc;
+using System.Threading.Tasks;
+
+namespace ExaLearn.WebApi.Controllers
+{
+    [Route("api/database")]
+    //[Authorize]
```
Comment (romanstat) [inline InternshipBE/ExaLearn.WebApi/Controllers/DatabaseController.cs]: Remove comment

File: InternshipBE/ExaLearn.WebApi/Startup.cs [hunk 17]
```diff
@@ -29,13 +34,57 @@ public Startup(IConfiguration configuration)
...
+                c.AddSecurityDefinition("Bearer", new OpenApiSecurityScheme
+                {
+                    In = ParameterLocation.Header,
+                    Description = "Please insert JWT with Bearer into field",
+                    Name = "Authorization",
+                    Type = SecuritySchemeType.ApiKey,
+                    BearerFormat = "JWT",
+                    Scheme = "Bearer"
+                });
+                c.AddSecurityRequirement(new OpenApiSecurityRequirement {
```
Comment (romanstat) [inline InternshipBE/ExaLearn.WebApi/Startup.cs]: curly brace to next line

**Review by romanstat — APPROVED** [review]

**Review by ExadelSandbox — APPROVED** [review]

**Pull request merged**